### DISCUSSION TITLE
Set Target Inclination doesn't work

### DISCUSSION THREAD

**Discussion by EkoFreshFTech** [hbldh/bleak#1623] [Q&A]
* bleak version: 0.22.2
* Python version: 3.11.9
* Operating System: Windows

### What I Did
I tried to give my hometrainer an inclination e.g 1%
I searched for the uuid of the characteristic and found Fitness Machine Control Point write (00002ad9-0000-1000-8000-00805f9b34fb)
Then i tried to request control so i can write the inclination i want. I used the Bluetooth FTMS Documenation.

```python
import asyncio
from bleak import BleakClient

CHARACTERISTIC_UUID = "00002ad9-0000-1000-8000-00805f9b34fb"
address = 'C5:B3:F0:FB:9A:1E' #Adress of my hometrainer
REQUEST_CONTROL_OPCODE = 0x00
SET_TARGET_INCLINATION_OPCODE = 0x03
TARGET_INCLINATION = 1  

async def main():
    async with BleakClient(address) as client:
        request_control = bytearray([REQUEST_CONTROL_OPCODE])
        await client.write_gatt_char(CHARACTERISTIC_UUID, request_control)
        print("Control requested")
        
        inclination_value = (TARGET_INCLINATION * 10).to_bytes(2, byteorder='little', signed=True)
        set_inclination_command = bytearray([SET_TARGET_INCLINATION_OPCODE]) + inclination_value
        await client.write_gatt_char(CHARACTERISTIC_UUID, set_inclination_command)
        print(f"Target inclination set to {TARGET_INCLINATION}%")

# Run the script
asyncio.run(main())
```

I always get this Error and i dont know how to solve it
```
Traceback (most recent call last):
  File "c:\Users\eekse\Desktop\Bachelorarbeit\Bluetooth\test.py", line 22, in <module>
    asyncio.run(main())
  File "C:\Users\eekse\AppData\Local\Programs\Python\Python311\Lib\asyncio\runners.py", line 190, in run
    return runner.run(main)
           ^^^^^^^^^^^^^^^^
  File "C:\Users\eekse\AppData\Local\Programs\Python\Python311\Lib\asyncio\runners.py", line 118, in run
    return self._loop.run_until_complete(task)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\eekse\AppData\Local\Programs\Python\Python311\Lib\asyncio\base_events.py", line 654, in run_until_complete
    return future.result()
           ^^^^^^^^^^^^^^^
  File "c:\Users\eekse\Desktop\Bachelorarbeit\Bluetooth\test.py", line 13, in main
    await client.write_gatt_char(CHARACTERISTIC_UUID, request_control)
  File "C:\Users\eekse\AppData\Local\Programs\Python\Python311\Lib\site-packages\bleak\__init__.py", line 786, in write_gatt_char
    await self._backend.write_gatt_char(characteristic, data, response)
  File "C:\Users\eekse\AppData\Local\Programs\Python\Python311\Lib\site-packages\bleak\backends\winrt\client.py", line 929, in write_gatt_char
    _ensure_success(
  File "C:\Users\eekse\AppData\Local\Programs\Python\Python311\Lib\site-packages\bleak\backends\winrt\client.py", line 160, in _ensure_success
    raise BleakError(
bleak.exc.BleakError: Could not write value bytearray(b'\x00') to characteristic 0031: Protocol Error 0x81: Unknown
```
I also tried to directly write the inclination without requesting control but I get the same error. I dont know how to solver this
I would very appreciate if someone can help me.

**Comment by dlech** [comment]
This error comes from the Windows Bluetooth stack, so I would suggest doing a Bluetooth packet capture to see if that can provide more insight. (see troubleshooting page in the docs for instructions).

Also, if you have the "official" app for this device, you can capture Bluetooth packets while the official app is running to see how that works, then try to reproduce that using Bleak.

**Reply by EkoFreshFTech** [reply]
I'm new to bluetooth data traffic or in general handling data. I captured some data from "Adapter for loopback capture" ( this was the only one changing when I run my script) from the start till the error pops up.  But I have no clue what to do with the information of the capture. E.g. the end of the capture 
![image](https://github.com/user-attachments/assets/7c061291-9452-41b8-b4da-c594cd7f2d02)
No idea what to do.

**Reply by dlech** [reply]
That is network traffic, not Bluetooth. See https://bleak.readthedocs.io/en/latest/troubleshooting.html#windows-10 for the instructions on how to get Bluetooth captures on Windows.

**Reply by EkoFreshFTech** [reply]
This should be the correct one. Screenshotted everything 
![image](https://github.com/user-attachments/assets/ad110313-6610-4fe1-bd66-e81dcdb080d1)
![image](https://github.com/user-attachments/assets/c655e976-f5a3-485e-8a62-b781d4cad854)
![image](https://github.com/user-attachments/assets/1a7ddad7-7528-4cce-b7d6-f509c0ebb149)
![image](https://github.com/user-attachments/assets/818599a2-f65a-4827-b09c-998ee77797d5)
![image](https://github.com/user-attachments/assets/f0c93e4f-3ec5-425d-9e7a-c1f42223bb0d)
![image](https://github.com/user-attachments/assets/4340e060-b659-434b-986e-0afd1bffba75)
![image](https://github.com/user-attachments/assets/1aa4f81c-aa44-4b9a-8a94-9a2491adac2e)
Here the file.
[ble_package.zip](https://github.com/user-attachments/files/16393464/ble_package.zip)

**Reply by dlech** [reply]
Towards the end, we can see the same error that we saw in Bleak.

<img width="1043" alt="image" src="https://github.com/user-attachments/assets/5b912b66-4504-492f-bd44-fa1c6890c523">

So we can tell the error is coming from the remote device. So apparently, the device doesn't like the message you are sending.

0x81 is in the application error range, which means it is specific to this device, which is why the error message says `0x81: Unknown`.

I haven't used this fitness service before, so I don't know how it is actually supposed to work.

If you have an app that can control the fitness machine and can do a Bluetooth capture while that app is running, we could look at it and try to reverse engineer it to see what we need to do to make it work with Bleak.

**Reply by EkoFreshFTech** [reply]
All right, thanks for all your Help :D Unfortunately I can't work on it in the weekend. Until Monday I do some research regarding the device. I already used an app to enable Fast track on the same device to increase the data sent to 10 Hz. I really appreciate your help 😄. Have a nice weekend.

**Reply by ukBaz** [reply]
@EkoFreshFTech ,
You might also want to try using a generic Bluetooth Low Energy scanning and exploration tool like [nRF Connect](https://www.nordicsemi.com/Products/Development-tools/nRF-Connect-for-mobile) to explore what characteristics are there and the process for changing values

I've also taken a look at the [Fitness Machine Service](https://www.bluetooth.com/specifications/specs/fitness-machine-service-1-0/) on the Bluetooth website. Each of the op-codes should give a response as to if they have been successful. Have you tried checking these response values?

Does changing of 

Here are some key chapters from the spec:

![image](https://github.com/user-attachments/assets/cbea43f0-ebd2-42c1-9306-4e7a9cb00b27)

![image](https://github.com/user-attachments/assets/c58b80ea-59cc-49bc-ab82-7da7f259f226)

![image](https://github.com/user-attachments/assets/a591d0bf-0dfb-42b4-9b44-19be1294a951)

![image](https://github.com/user-attachments/assets/b22b9148-d073-4177-84ec-06c6e8233550)

![image](https://github.com/user-attachments/assets/ae141741-7fab-4918-bcfd-5a70d95885ab)
